File: service-examples/rust/echo/src/bin/golden_test.rs
//! Golden file deserialization test (Rust).
//! Verifies Rust can deserialize the golden msgpack files.
//! Usage: golden_test --golden-dir golden/

use echo_wire_compat::types_gen::*;
use std::fs;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let golden_dir = args.iter()
        .position(|a| a == "--golden-dir")
        .and_then(|i| args.get(i + 1))
        .expect("Usage: golden_test --golden-dir <path>");

    let mut pass = 0;
    let mut fail = 0;

    // Request golden files
    match check_request::<EchoBytes>(golden_dir, "echo_bytes_request.msgpack") {
        Ok(cmd) => {
            assert_eq!(cmd.data, vec![0xDE, 0xAD, 0xBE, 0xEF, 0x42]);
            eprintln!("  PASS: echo_bytes_request.msgpack");
            pass += 1;
        }
        Err(e) => { eprintln!("  FAIL: echo_bytes_request.msgpack: {e}"); fail += 1; }
    }

    match check_request::<EchoFields>(golden_dir, "echo_fields_request.msgpack") {
        Ok(cmd) => {
            assert_eq!(cmd.a, 42);
            assert_eq!(cmd.b, 999999);
            assert_eq!(cmd.name, "hello wire compat");
            eprintln!("  PASS: echo_fields_request.msgpack");
            pass += 1;
        }
        Err(e) => { eprintln!("  FAIL: echo_fields_request.msgpack: {e}"); fail += 1; }
    }

    match check_request::<EchoNested>(golden_dir, "echo_nested_request.msgpack") {
        Ok(cmd) => {
            assert_eq!(cmd.inner.values, vec![vec![1u8, 2, 3], vec![4, 5]]);
            assert_eq!(cmd.inner.flag, Some(true));
            eprintln!("  PASS: echo_nested_request.msgpack");
            pass += 1;
        }
        Err(e) => { eprintln!("  FAIL: echo_nested_request.msgpack: {e}"); fail += 1; }
    }

    // Response golden files
    match check_response(golden_dir, "echo_bytes_response.msgpack") {
        Ok(Response::EchoBytesResponse(r)) => {
            assert_eq!(r.data, vec![0xDE, 0xAD, 0xBE, 0xEF, 0x42]);
            eprintln!("  PASS: echo_bytes_response.msgpack");
            pass += 1;
        }
        Ok(_) => { eprintln!("  FAIL: echo_bytes_response.msgpack: wrong variant"); fail += 1; }
        Err(e) => { eprintln!("  FAIL: echo_bytes_response.msgpack: {e}"); fail += 1; }
    }

    match check_response(golden_dir, "echo_fields_response.msgpack") {
        Ok(Response::EchoFieldsResponse(r)) => {
            assert_eq!(r.a, 42);
            assert_eq!(r.b, 999999);
            assert_eq!(r.name, "hello wire compat");
            eprintln!("  PASS: echo_fields_response.msgpack");
            pass += 1;
        }
        Ok(_) => { eprintln!("  FAIL: echo_fields_response.msgpack: wrong variant"); fail += 1; }
        Err(e) => { eprintln!("  FAIL: echo_fields_response.msgpack: {e}"); fail += 1; }
    }

    match check_response(golden_dir, "echo_nested_response.msgpack") {
        Ok(Response::EchoNestedResponse(r)) => {
            assert_eq!(r.inner.values, vec![vec![1u8, 2, 3], vec![4, 5]]);
            assert_eq!(r.inner.flag, Some(true));
            eprintln!("  PASS: echo_nested_response.msgpack");
            pass += 1;
        }
        Ok(_) => { eprintln!("  FAIL: echo_nested_response.msgpack: wrong variant"); fail += 1; }
        Err(e) => { eprintln!("  FAIL: echo_nested_response.msgpack: {e}"); fail += 1; }
    }

    eprintln!("\nResults: {pass}/{} passed, {fail} failed", pass + fail);
    if fail > 0 { std::process::exit(1); }
}

fn check_request<T: serde::de::DeserializeOwned>(dir: &str, name: &str) -> Result<T, String> {
    let path = format!("{dir}/{name}");
    let data = fs::read(&path).map_err(|e| format!("read {path}: {e}"))?;
    // Request format: [Command] — deserialized as Vec<Command>
    let commands: Vec<Command> = rmp_serde::from_slice(&data)
        .map_err(|e| format!("deserialize: {e}"))?;
    let command = commands.into_iter().next().ok_or("empty")?;
    // Extract the specific variant
    // We need to serialize the inner value back to msgpack and then deserialize as T
    let inner_bytes = match command {
        Command::EchoBytes(v) => rmp_serde::to_vec_named(&v).unwrap(),
        Command::EchoFields(v) => rmp_serde::to_vec_named(&v).unwrap(),
        Command::EchoNested(v) => rmp_serde::to_vec_named(&v).unwrap(),
        Command::EchoShutdown(v) => rmp_serde::to_vec_named(&v).unwrap(),
    };
    rmp_serde::from_slice(&inner_bytes).map_err(|e| format!("re-deserialize: {e}"))
}

fn check_response(dir: &str, name: &str) -> Result<Response, String> {
    let path = format!("{dir}/{name}");
    let data = fs::read(&path).map_err(|e| format!("read {path}: {e}"))?;
    rmp_serde::from_slice(&data).map_err(|e| format!("deserialize: {e}"))
}

File: ipc-codegen/examples/ts/echo/golden_test.ts
/**
 * Golden file wire compatibility test (TypeScript).
 *
 * Verifies that TypeScript can correctly deserialize msgpack data produced by
 * the Rust reference implementation (the golden files). This is the critical
 * cross-language compatibility check — if TS can read Rust's output, and the
 * round-trip tests show Rust can read TS's output, wire compat is proven.
 *
 * Usage: npx tsx golden_test.ts
 * Exits 0 if all pass, 1 on failure.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { Decoder } from 'msgpackr';

const decoder = new Decoder({ useRecords: false });
const goldenDir = path.join(import.meta.dirname!, '../../echo-schema', 'golden');

let pass = 0;
let fail = 0;

function assertEqual(actual: any, expected: any, label: string) {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  if (a !== e) {
    throw new Error(`${label}: expected ${e}, got ${a}`);
  }
}

function bufEqual(a: Uint8Array, b: number[]) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function checkGoldenRequest(name: string, expectedCmdName: string, validate: (fields: any) => void) {
  try {
    const golden = fs.readFileSync(path.join(goldenDir, name));
    const decoded = decoder.unpack(golden) as any[];
    // Request format: [[commandName, {fields}]]
    assertEqual(decoded.length, 1, `${name} array length`);
    const [cmdName, fields] = decoded[0];
    assertEqual(cmdName, expectedCmdName, `${name} command name`);
    validate(fields);
    console.log(`  PASS: ${name}`);
    pass++;
  } catch (e: any) {
    console.log(`  FAIL: ${name}: ${e.message}`);
    fail++;
  }
}

function checkGoldenResponse(name: string, expectedRespName: string, validate: (fields: any) => void) {
  try {
    const golden = fs.readFileSync(path.join(goldenDir, name));
    const decoded = decoder.unpack(golden) as any[];
    // Response format: [responseName, {fields}]
    assertEqual(decoded.length, 2, `${name} array length`);
    const [respName, fields] = decoded;
    assertEqual(respName, expectedRespName, `${name} response name`);
    validate(fields);
    console.log(`  PASS: ${name}`);
    pass++;
  } catch (e: any) {
    console.log(`  FAIL: ${name}: ${e.message}`);
    fail++;
  }
}

console.log('Golden file deserialization tests (TypeScript):\n');

// Request golden files
checkGoldenRequest('echo_bytes_request.msgpack', 'EchoBytes', (f) => {
  if (!bufEqual(f.data, [0xDE, 0xAD, 0xBE, 0xEF, 0x42])) {
    throw new Error('data mismatch');
  }
});

checkGoldenRequest('echo_fields_request.msgpack', 'EchoFields', (f) => {
  assertEqual(f.a, 42, 'a');
  assertEqual(f.b, 999999, 'b');
  assertEqual(f.name, 'hello wire compat', 'name');
});

checkGoldenRequest('echo_nested_request.msgpack', 'EchoNested', (f) => {
  assertEqual(f.inner.flag, true, 'flag');
  assertEqual(f.inner.values.length, 2, 'values length');
  if (!bufEqual(f.inner.values[0], [1, 2, 3])) throw new Error('values[0] mismatch');
  if (!bufEqual(f.inner.values[1], [4, 5])) throw new Error('values[1] mismatch');
});

// Response golden files
checkGoldenResponse('echo_bytes_response.msgpack', 'EchoBytesResponse', (f) => {
  if (!bufEqual(f.data, [0xDE, 0xAD, 0xBE, 0xEF, 0x42])) {
    throw new Error('data mismatch');
  }
});

checkGoldenResponse('echo_fields_response.msgpack', 'EchoFieldsResponse', (f) => {
  assertEqual(f.a, 42, 'a');
  assertEqual(f.b, 999999, 'b');
  assertEqual(f.name, 'hello wire compat', 'name');
});

checkGoldenResponse('echo_nested_response.msgpack', 'EchoNestedResponse', (f) => {
  assertEqual(f.inner.flag, true, 'flag');
  assertEqual(f.inner.values.length, 2, 'values length');
  if (!bufEqual(f.inner.values[0], [1, 2, 3])) throw new Error('values[0] mismatch');
  if (!bufEqual(f.inner.values[1], [4, 5])) throw new Error('values[1] mismatch');
});

console.log(`\nResults: ${pass}/${pass + fail} passed, ${fail} failed`);
if (fail > 0) process.exit(1);
